File: ios/FamiliarSC/Sources/FamiliarSC/Autonomy.swift
import Foundation

// The autonomy dial, read exactly as crates/whisker/src/autonomy.rs writes and reads it —
// Ian's ruling (2026-09-03): advise / confirm / auto, per control-surface category and
// family. The app SHOWS the dial and lets the captain change it; the change is the captain's
// act on their tap (the app writes autonomy.json itself), never a model's. This file keeps
// the precedence and the vocabulary in lockstep with the Rust side — the tests pin both.

public enum AutonomyLevel: String, Codable, CaseIterable, Equatable, Sendable {
    case advise, confirm, auto

    /// Mirrors `Level::parse` — the same aliases a captain may type.
    public static func parse(_ s: String) -> AutonomyLevel? {
        switch s.trimmingCharacters(in: .whitespaces).lowercased() {
        case "advise", "advice", "advisory": return .advise
        case "confirm", "ask": return .confirm
        case "auto", "automatic", "autonomous": return .auto
        default: return nil
        }
    }
}

/// A control surface, `family.category` — the pilot's doors today, plus racing's.
public enum ControlSurface: String, CaseIterable, Equatable, Sendable {
    case navigationCourse = "navigation.course"
    case navigationFuel = "navigation.fuel"
    case navigationRescue = "navigation.rescue"
    case freightBook = "freight.book"
    case freightCollect = "freight.collect"
    case freightCancel = "freight.cancel"
    case marketBuy = "market.buy"
    case marketSell = "market.sell"
    case marketCarry = "market.carry"
    /// Trading on the credit line (Ian, 2026-09-05: borrowing to speculate is a dial the
    /// captain owns). Advise by default, like the tanker: an unconfigured captain is
    /// OFFERED the borrow. Missing here until 2026-09-08 — a valid host file carrying it
    /// read as malformed (codex T-237 B2 re-verification, finding 2).
    case marketMargin = "market.margin"
    case shipRepair = "ship.repair"
    case shipRefit = "ship.refit"
    case shipCrew = "ship.crew"
    case shipFrame = "ship.frame"
    case shipLease = "ship.lease"
    case racingPlot = "racing.plot"
    case racingLine = "racing.line"
    case racingRefusal = "racing.refusal"

    public static let families = ["navigation", "freight", "market", "ship", "racing"]

    public var key: String { rawValue }
    public var family: String { String(rawValue.split(separator: ".")[0]) }
    public var category: String { String(rawValue.split(separator: ".")[1]) }

    public static func parse(_ s: String) -> ControlSurface? {
        ControlSurface(rawValue: s.trimmingCharacters(in: .whitespaces))
    }

    /// The level an unconfigured captain gets — `Dial::level`'s `unwrap_or` arm.
    public var unconfiguredDefault: AutonomyLevel {
        switch self {
        case .navigationRescue, .marketMargin: return .advise
        default: return .auto
        }
    }

    /// Which automation grant a surface belongs to — a surface whose automation the captain
    /// has not bought is not on the dial at all (the grant model beneath it).
    public var automation: String? {
        switch family {
        case "navigation", "freight": return "freight"
        case "market": return "trade"
        case "ship": return "outfit"
        default: return nil
        }
    }
}

/// The dial as the store holds it: `{"navigation.course": "auto", "market": "confirm", "*": "auto"}`.
public struct AutonomyDial: Equatable, Sendable {
    public var settings: [String: AutonomyLevel]

    public init(settings: [String: AutonomyLevel] = [:]) { self.settings = settings }

    /// Most specific wins: category, then family, then `*`, then the default — auto for
    /// everything bought, except the two the captain must be asked about first: the tanker
    /// (`navigation.rescue`, a PAWS call is a multi-day strand that pins the hull) and the
    /// credit line (`market.margin`). The defaults are pinned against
    /// `Fixtures/contract/autonomy-surfaces.json`, which the Rust side pins too.
    public func level(for s: ControlSurface) -> AutonomyLevel {
        if let l = settings[s.key] { return l }
        if let l = settings[s.family] { return l }
        if let l = settings["*"] { return l }
        return s.unconfiguredDefault
    }

    /// Mirrors `Dial::set`: `*`, a family, or a surface; anything else is refused.
    public mutating func set(_ key: String, _ level: AutonomyLevel) -> String? {
        let k = key.trimmingCharacters(in: .whitespaces)
        let known = k == "*" || ControlSurface.families.contains(k) || ControlSurface.parse(k) != nil
        guard known else { return "unknown control surface `\(k)`" }
        settings[k] = level
        return nil
    }

    /// Pretty JSON in the file's own shape (a flat object, keys sorted).
    public func encoded() -> Data {
        let obj = Dictionary(uniqueKeysWithValues: settings.map { ($0.key, $0.value.rawValue) })
        let enc = JSONEncoder()
        enc.outputFormatting = [.prettyPrinted, .sortedKeys]
        return (try? enc.encode(obj)) ?? Data("{}".utf8)
    }

    public static func decode(_ data: Data) throws -> AutonomyDial {
        let raw = try JSONCoding.decode([String: String].self, from: data)
        var dial = AutonomyDial()
        for (k, v) in raw {
            guard let l = AutonomyLevel(rawValue: v) else {
                throw StoreError.invalid("autonomy.json", "`\(k)`: \"\(v)\" is not advise|confirm|auto")
            }
            if let why = dial.set(k, l) { throw StoreError.invalid("autonomy.json", why) }
        }
        return dial
    }

    /// What whisker's own loader does with the file — and what the captain must be told.
    public enum Loaded: Equatable, Sendable {
        /// No file: every bought surface is auto (KK II today).
        case absent
        case dial(AutonomyDial)
        /// The file is there but whisker cannot parse it — and whisker's loader then
        /// treats it as ABSENT, i.e. auto everywhere. The app must say so, loudly.
        case malformed(String)

        public var dial: AutonomyDial {
            if case .dial(let d) = self { return d }
            return AutonomyDial()
        }
    }

    public static func load(from url: URL) -> Loaded {
        guard FileManager.default.fileExists(atPath: url.path) else { return .absent }
        guard let data = try? Data(contentsOf: url) else { return .malformed("unreadable") }
        do { return .dial(try decode(data)) } catch let e as StoreError { return .malformed(e.description) } catch {
            return .malformed("not JSON")
        }
    }
}

File: crates/pilot/src/autonomy.rs
//! The autonomy dial, pure: how much of each control surface the captain has handed
//! to the ship's computer. The asked-for shape (2026-09-03): a message window giving
//! advice the captain may follow or not; the ability to simply allow a captain-authorized
//! command and have the ship proceed as KK II already does; and granular configuration of
//! autonomy per control-surface category and family.
//!
//! Every doctrine decision has one of three fates, chosen per control surface:
//! **advise** (say what it would do, do nothing), **confirm** (propose, act only on the
//! captain's yes within a fold), **auto** (act and journal). The dial is `autonomy.json`
//! in the ship store: `{"navigation.course": "auto", "market": "confirm", "*": "auto"}` —
//! a category, a family, or `*`, most specific wins. Absent means auto for everything
//! the captain has bought (KK II today). The grant model stands beneath it: an
//! automation not bought is not on the dial at all.

use std::collections::BTreeMap;

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum Level {
    Advise,
    Confirm,
    Auto,
}

impl Level {
    pub fn parse(s: &str) -> Option<Self> {
        match s.trim().to_ascii_lowercase().as_str() {
            "advise" | "advice" | "advisory" => Some(Self::Advise),
            "confirm" | "ask" => Some(Self::Confirm),
            "auto" | "automatic" | "autonomous" => Some(Self::Auto),
            _ => None,
        }
    }
    pub fn name(self) -> &'static str {
        match self {
            Level::Advise => "advise",
            Level::Confirm => "confirm",
            Level::Auto => "auto",
        }
    }
}

/// A control surface: `family.category`. The families and categories the pilot
/// exercises today, plus racing's for when it lands.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
pub enum Surface {
    NavigationCourse,
    NavigationFuel,
    NavigationRescue,
    FreightBook,
    FreightCollect,
    FreightCancel,
    MarketBuy,
    MarketSell,
    MarketCarry,
    MarketMargin,
    ShipRepair,
    ShipRefit,
    ShipCrew,
    ShipFrame,
    ShipLease,
    RacingPlot,
    RacingLine,
    RacingRefusal,
}

impl Surface {
    pub fn family(self) -> &'static str {
        match self {
            Surface::NavigationCourse | Surface::NavigationFuel | Surface::NavigationRescue => {
                "navigation"
            }
            Surface::FreightBook | Surface::FreightCollect | Surface::FreightCancel => "freight",
            Surface::MarketBuy
            | Surface::MarketSell
            | Surface::MarketCarry
            | Surface::MarketMargin => "market",
            Surface::ShipRepair
            | Surface::ShipRefit
            | Surface::ShipCrew
            | Surface::ShipFrame
            | Surface::ShipLease => "ship",
            Surface::RacingPlot | Surface::RacingLine | Surface::RacingRefusal => "racing",
        }
    }
    pub fn category(self) -> &'static str {
        match self {
            Surface::NavigationCourse => "course",
            Surface::NavigationFuel => "fuel",
            Surface::NavigationRescue => "rescue",
            Surface::FreightBook => "book",
            Surface::FreightCollect => "collect",
            Surface::FreightCancel => "cancel",
            Surface::MarketBuy => "buy",
            Surface::MarketSell => "sell",
            Surface::MarketCarry => "carry",
            Surface::MarketMargin => "margin",
            Surface::ShipRepair => "repair",
            Surface::ShipRefit => "refit",
            Surface::ShipCrew => "crew",
            Surface::ShipFrame => "frame",
            Surface::ShipLease => "lease",
            Surface::RacingPlot => "plot",
            Surface::RacingLine => "line",
            Surface::RacingRefusal => "refusal",
        }
    }
    pub fn key(self) -> String {
        format!("{}.{}", self.family(), self.category())
    }
    pub fn all() -> &'static [Surface] {
        &[
            Surface::NavigationCourse,
            Surface::NavigationFuel,
            Surface::NavigationRescue,
            Surface::FreightBook,
            Surface::FreightCollect,
            Surface::FreightCancel,
            Surface::MarketBuy,
            Surface::MarketSell,
            Surface::MarketCarry,
            Surface::MarketMargin,
            Surface::ShipRepair,
            Surface::ShipRefit,
            Surface::ShipCrew,
            Surface::ShipFrame,
            Surface::ShipLease,
            Surface::RacingPlot,
            Surface::RacingLine,
            Surface::RacingRefusal,
        ]
    }
    pub fn parse(s: &str) -> Option<Surface> {
        Surface::all().iter().copied().find(|x| x.key() == s.trim())
    }
}

/// The dial as the store holds it: keys are `family.category`, `family`, or `*`.
#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct Dial {
    #[serde(flatten)]
    pub settings: BTreeMap<String, Level>,
}

pub const DIAL_FILE: &str = "autonomy.json";

impl Dial {
    /// The dial from its stored JSON. Unreadable or absent text is the default dial —
    /// what cannot be read grants nothing beyond the floor. (The file itself is the
    /// runner's: `store::load_dial`. This crate owns no file.)
    pub fn parse(text: &str) -> Dial {
        serde_json::from_str(text).unwrap_or_default()
    }
    /// The dial as the store writes it.
    pub fn to_json(&self) -> Vec<u8> {
        serde_json::to_vec_pretty(self).unwrap_or_default()
    }
    /// Most specific setting wins: category, then family, then `*`, then auto. The
    /// one surface with its own floor is the tanker on a real-time world — `rescue`
    /// defaults to ADVISE unless the captain says otherwise, because a PAWS call is
    /// a multi-day strand that also pins the hull (learned on Kibble Klipper II, 2026-09).
    pub fn level(&self, s: Surface) -> Level {
        self.settings
            .get(&s.key())
            .or_else(|| self.settings.get(s.family()))
            .or_else(|| self.settings.get("*"))
            .copied()
            .unwrap_or(match s {
                // The tanker on a real-time world: a PAWS call is a multi-day
                // strand that also pins the hull (learned on Kibble Klipper II,
                // 2026-09).
                Surface::NavigationRescue => Level::Advise,
                // Trading on the credit line. The owner's ruling (2026-09-05):
                // the automation should include the ability to borrow to
                // speculate, and that should be a dial in the bridge's autonomy
                // settings — working capital should be utilized to maximize the
                // captain's success. So the ability exists and the captain owns
                // it; an unconfigured captain is
                // OFFERED the borrow rather than committed to it, the same
                // reasoning as the tanker. A captain who has set `*` has already
                // spoken and gets what they asked for.
                Surface::MarketMargin => Level::Advise,
                _ => Level::Auto,
            })
    }
    pub fn set(&mut self, key: &str, level: Level) -> Result<(), String> {
        let k = key.trim();
        let known = k == "*"
            || ["navigation", "freight", "market", "ship", "racing"].contains(&k)
            || Surface::parse(k).is_some();
        if !known {
            return Err(format!("unknown control surface `{k}`"));
        }
        self.settings.insert(k.to_string(), level);
        Ok(())
    }
}

/// A proposed act awaiting the captain (`proposals.jsonl`), and the captain's word
/// on it (`approvals.jsonl`: `{"id": …, "approved": true|false, "at": …}`).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Proposal {
    pub id: String,
    pub tick: i64,
    /// The last tick this proposal may be acted on; after it, it lapses.
    pub expires_tick: i64,
    pub surface: String,
    pub describe: String,
    pub why: String,
    pub body: serde_json::Value,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Approval {
    pub id: String,
    pub approved: bool,
    pub at: i64,
}

/// What the gate says about acting on a decision now.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Gate {
    /// Act, and journal it.
    Act,
    /// Say what you would do; do not act.
    Advise,
    /// Proposed and waiting; act on a later fold if approved.
    Proposed,
    /// Proposed earlier, lapsed unapproved.
    Lapsed,
}

/// A proposal is identified by its surface and body, so the same intent on the next
/// fold finds its own earlier proposal rather than filing another.
pub fn proposal_id(surface: Surface, body: &serde_json::Value) -> String {
    let mut h: u64 = 0xcbf29ce484222325;
    for b in format!("{}|{}", surface.key(), body).bytes() {
        h ^= b as u64;
        h = h.wrapping_mul(0x100000001b3);
    }
    format!("p-{h:016x}")
}

/// The gate: given the dial, the surface, the tick, the fold length for the TTL, the
/// proposals already on file and the approvals, decide this fold's fate for the act.
/// `now_proposal` is filled in when a new proposal must be written.
#[allow(clippy::too_many_arguments)]
pub fn gate(
    dial: &Dial,
    surface: Surface,
    tick: i64,
    body: &serde_json::Value,
    describe: &str,
    why: &str,
    proposals: &[Proposal],
    approvals: &[Approval],
    new_proposal: &mut Option<Proposal>,
) -> Gate {
    match dial.level(surface) {
        Level::Auto => Gate::Act,
        Level::Advise => Gate::Advise,
        Level::Confirm => {
            let id = proposal_id(surface, body);
            if let Some(a) = approvals.iter().rev().find(|a| a.id == id) {
                return if a.approved { Gate::Act } else { Gate::Lapsed };
            }
            match proposals.iter().rev().find(|p| p.id == id) {
                Some(p) if tick <= p.expires_tick => Gate::Proposed,
                Some(_) => Gate::Lapsed,
                None => {
                    *new_proposal = Some(Proposal {
                        id,
                        tick,
                        // One fold of grace beyond the next: enough for a captain
                        // watching, not enough for a stale plan to fire later.
                        expires_tick: tick + PROPOSAL_TTL_TICKS,
                        surface: surface.key(),
                        describe: describe.to_string(),
                        why: why.to_string(),
                        body: body.clone(),
                    });
                    Gate::Proposed
                }
            }
        }
    }
}

/// How long a proposal waits for the captain before it lapses, in ticks.
pub const PROPOSAL_TTL_TICKS: i64 = 4;

/// Proposals from their jsonl text, torn lines skipped. (Files are the runner's:
/// `store::load_proposals`.)
pub fn parse_proposals(text: &str) -> Vec<Proposal> {
    text.lines()
        .filter_map(|l| serde_json::from_str(l).ok())
        .collect()
}

/// Approvals from their jsonl text, torn lines skipped.
pub fn parse_approvals(text: &str) -> Vec<Approval> {
    text.lines()
        .filter_map(|l| serde_json::from_str(l).ok())
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn absent_dial_is_auto_for_everything_but_the_tanker() {
        let d = Dial::default();
        assert_eq!(d.level(Surface::FreightBook), Level::Auto);
        assert_eq!(d.level(Surface::MarketBuy), Level::Auto);
        assert_eq!(d.level(Surface::NavigationRescue), Level::Advise);
    }

    #[test]
    fn most_specific_setting_wins() {
        let mut d = Dial::default();
        d.set("*", Level::Advise).unwrap();
        d.set("market", Level::Confirm).unwrap();
        d.set("market.sell", Level::Auto).unwrap();
        assert_eq!(d.level(Surface::FreightBook), Level::Advise);
        assert_eq!(d.level(Surface::MarketBuy), Level::Confirm);
        assert_eq!(d.level(Surface::MarketSell), Level::Auto);
        assert!(d.set("kitchen.sink", Level::Auto).is_err());
        let json = serde_json::to_string(&d).unwrap();
        let back: Dial = serde_json::from_str(&json).unwrap();
        assert_eq!(back, d);
    }

    #[test]
    fn confirm_proposes_once_acts_on_approval_and_lapses_unapproved() {
        let mut d = Dial::default();
        d.set("freight.book", Level::Confirm).unwrap();
        let body = json!({"type": "book", "loadId": "L1"});
        let mut fresh = None;
        let g = gate(
            &d,
            Surface::FreightBook,
            100,
            &body,
            "book L1",
            "best rate",
            &[],
            &[],
            &mut fresh,
        );
        assert_eq!(g, Gate::Proposed);
        let p = fresh.clone().expect("a proposal was written");
        assert_eq!(p.expires_tick, 100 + PROPOSAL_TTL_TICKS);
        // Next fold, still waiting: the same proposal, not a second one.
        let mut again = None;
        let g = gate(
            &d,
            Surface::FreightBook,
            101,
            &body,
            "book L1",
            "best rate",
            std::slice::from_ref(&p),
            &[],
            &mut again,
        );
        assert_eq!(g, Gate::Proposed);
        assert!(again.is_none());
        // The captain says yes.
        let yes = Approval {
            id: p.id.clone(),
            approved: true,
            at: 1,
        };
        let g = gate(
            &d,
            Surface::FreightBook,
            102,
            &body,
            "book L1",
            "best rate",
            std::slice::from_ref(&p),
            &[yes],
            &mut None,
        );
        assert_eq!(g, Gate::Act);
        // Or says nothing and the fold passes.
        let g = gate(
            &d,
            Surface::FreightBook,
            100 + PROPOSAL_TTL_TICKS + 1,
            &body,
            "book L1",
            "best rate",
            std::slice::from_ref(&p),
            &[],
            &mut None,
        );
        assert_eq!(g, Gate::Lapsed);
        // Or says no.
        let no = Approval {
            id: p.id.clone(),
            approved: false,
            at: 2,
        };
        let g = gate(
            &d,
            Surface::FreightBook,
            102,
            &body,
            "book L1",
            "best rate",
            &[p],
            &[no],
            &mut None,
        );
        assert_eq!(g, Gate::Lapsed);
    }

    #[test]
    fn advise_never_acts_and_auto_always_does() {
        let mut d = Dial::default();
        d.set("navigation", Level::Advise).unwrap();
        let body = json!({"type": "travel", "station": "foxys-diner"});
        assert_eq!(
            gate(
                &d,
                Surface::NavigationCourse,
                1,
                &body,
                "",
                "",
                &[],
                &[],
                &mut None
            ),
            Gate::Advise
        );
        assert_eq!(
            gate(
                &d,
                Surface::ShipRepair,
                1,
                &json!({"type":"repair"}),
                "",
                "",
                &[],
                &[],
                &mut None
            ),
            Gate::Act
        );
    }
}

/// The dial's vocabulary is a CONTRACT with any companion app that renders it, and it drifted
/// once: `market.margin` was added here and an app read a valid host file as malformed (a
/// design review finding). One fixture now names every surface and its unconfigured default,
/// and both sides pin that one file, so a surface added on either side fails the other until
/// both carry it.
#[cfg(test)]
mod contract_pins {
    use super::*;

    const CONTRACT: &str = include_str!("../tests/fixtures/contract/autonomy-surfaces.json");

    #[test]
    fn every_surface_and_its_default_match_the_shared_contract() {
        let c: serde_json::Value = serde_json::from_str(CONTRACT).expect("contract parses");
        let surfaces = c["surfaces"].as_object().expect("surfaces object");
        let dial = Dial::default();
        for &s in Surface::all() {
            let key = s.key();
            let want = surfaces
                .get(&key)
                .and_then(|v| v.as_str())
                .unwrap_or_else(|| panic!("surface {key} is not in the shared contract"));
            assert_eq!(dial.level(s).name(), want, "default level for {key}");
        }
        assert_eq!(
            surfaces.len(),
            Surface::all().len(),
            "the contract names a surface Rust lacks"
        );
        let families: Vec<&str> = c["families"]
            .as_array()
            .expect("families")
            .iter()
            .filter_map(|v| v.as_str())
            .collect();
        for f in &families {
            assert!(
                Surface::all().iter().any(|s| s.family() == *f),
                "contract family {f} has no surface"
            );
        }
        for &s in Surface::all() {
            assert!(
                families.contains(&s.family()),
                "family {} missing from the contract",
                s.family()
            );
        }
    }
}
